src/gz/bufread.rs

```diff
@@ -34,87 +34,107 @@ fn bad_header() -> io::Error {
-    let mtime = ((header[4] as u32) << 0)
-        | ((header[5] as u32) << 8)
-        | ((header[6] as u32) << 16)
-        | ((header[7] as u32) << 24);
-    let _xfl = header[8];
-    let os = header[9];
-
-    let extra = if flg & FEXTRA != 0 {
-        let xlen = read_le_u16(&mut crc_reader)?;
-        let mut extra = vec![0; xlen as usize];
-        crc_reader.read_exact(&mut extra)?;
-        Some(extra)
-    } else {
-        None
-    };
-    let filename = if flg & FNAME != 0 {
-        // wow this is slow
-        let mut b = Vec::new();
-        for byte in crc_reader.by_ref().bytes() {
-            let byte = byte?;
-            if byte == 0 {
-                break;
+fn read_gz_header_part<R: Read>(r: &mut Buffer<R>) -> io::Result<()> {
+    loop {
+        match r.part.state {
+            GzHeaderParsingState::Start => {
+                let mut header = [0; 10];
+                r.read_once(&mut header)?;
+                r.part.buf.truncate(0);
+
+                if header[0] != 0x1f || header[1] != 0x8b {
+                    return Err(bad_header());
+                }
+                if header[2] != 8 {
+                    return Err(bad_header());
+                }
+
+                r.part.flg = header[3];
+                r.part.header.mtime = ((header[4] as u32) << 0)
+                    | ((header[5] as u32) << 8)
+                    | ((header[6] as u32) << 16)
+                    | ((header[7] as u32) << 24);
+                let _xfl = header[8];
+                r.part.header.operating_system = header[9];
+                r.part.state = GzHeaderParsingState::Xlen;
             }
-            b.push(byte);
-        }
-        Some(b)
-    } else {
-        None
-    };
-    let comment = if flg & FCOMMENT != 0 {
-        // wow this is slow
-        let mut b = Vec::new();
-        for byte in crc_reader.by_ref().bytes() {
-            let byte = byte?;
-            if byte == 0 {
-                break;
+            GzHeaderParsingState::Xlen => {
+                if r.part.flg & FEXTRA != 0 {
+                    r.part.xlen = read_le_u16_once(r)?;
+                }
+                r.part.state = GzHeaderParsingState::Extra;
+            }
+            GzHeaderParsingState::Extra => {
+                if r.part.flg & FEXTRA != 0 {
+                    let mut extra = vec![0; r.part.xlen as usize];
+                    r.read_once(&mut extra)?;
+                    r.part.header.extra = Some(extra);
+                }
+                r.part.state = GzHeaderParsingState::Filename;
+            }
+            GzHeaderParsingState::Filename => {
+                if r.part.flg & FNAME != 0 {
+                    if None == r.part.header.filename {
+                        let vec = Vec::new();
+                        r.part.header.filename = Some(vec);
+                    };
+                    if let Some(ref mut b) = r.part.header.filename {
+                        for byte in r.reader.bytes() {
```

Comment:
I don't think that reaching into the `reader` here is also quite right since it bypasses the buffering that's otherwise happening in `Buffer`?

Reply:
I also think this is bad style even if it is intentional as the buffering takes directly place in the `GzHeader`
Should I rather add a method `read_until` to the `Buffer` ?

Reply:
I don't have a shape of code in mind, but there's a lot of confusing buffer-this-sometimes-dont logic going on here and there's very little documentation. I think something needs to be improved (code structure or similar) to make this more understandable.

Reply:
I agree that this is not nice.
I just pushed something new.
The difficulty is `r.bytes()` is a mutable borrow and we want another mutable borrow for `r.part.header.filename`
I did not manage to have a mutable reference to `r.part.header.filename` in `Buffer` because of lifetime issues such as 
```
error[E0623]: lifetime mismatch
  --> src/gz/bufread.rs:89:39
   |
46 | fn read_gz_header_part<R: Read>(r: &mut Buffer<R>) -> io::Result<()> {
   |                                    --------------
   |                                    |
   |                                    these two types are declared with different lifetimes...
...
89 |                     r.direct_buffer = r.part.header.filename.as_ref();
   |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ...but data from `r` flows into `r` here
```

So, in the end `Buffer::read` looks at the `GzHeaderParsingState` to decide where to buffer the data

Another (not so nice) pattern could be
```
                    r.direct = true;
                    let mut bytebuf = [0; 1];
                    loop {
                        r.read_exact(&mut bytebuf)?;
                        if bytebuf[0] == 0 {
                            r.direct = false;
                            break;
                        }
                        if let Some(ref mut b) = r.part.header.filename {
                            b.push(bytebuf[0]);
                        }
                    }
```